tests/many_to_one/tests.py

```diff
@@ -667,3 +667,35 @@ def test_add_then_remove_after_prefetch(self):
         self.assertEqual(city.districts.count(), 2)
         city.districts.remove(d2)
         self.assertEqual(city.districts.count(), 1)
+
+    def test_cannot_reuse_trimmed_joins(self):
+        """
+        When a trimmable join is specified in the query (here city__), the ORM
+        detects it and removes unnecessary joins. The set of reusable joins
+        should be updated after trimming the query, so that other lookups
+        doesn't consider that outer query's filters are in effect for the
+        subquery.
+
+        See #26551 for more details.
+        """
+        vancouver = City.objects.create(id=1, name="Vancouver")
+        paris = City.objects.create(id=2, name="Paris")
+        Road.objects.create(city=vancouver, name="6th")
+        r = Road.objects.create(city=paris, name="10th")
+        District.objects.create(city=paris, name="XVII")
+        District.objects.create(city=vancouver, name="Fairview")
+
+        qs = Road.objects.filter(
+            # many_to_one_city is trimmable, do not add it to reusable set
+            ~(models.Q(city__districts__name='Fairview') &
+              # If many_to_one_city is in the reusable set, split_exclude will
```

Comment:
I'm not sure about including lots of details of the implementation in the test, considering that if the code is later refactored, someone is unlikely to also update comments. I suppose they may be helpful for reviewing the patch.